components/raftstore/src/store/util.rs

```diff
@@ -785,6 +800,88 @@ impl PerfContextStatistics {
     }
 }
 
+#[derive(PartialEq, Eq, Debug)]
+pub enum ConfChangeKind {
+    // Only contains one configuration change
+    Simple,
+    // Enter joint state
+    EnterJoint,
+    // Leave joint state
+    LeaveJoint,
+}
+
+impl ConfChangeKind {
+    pub fn confchange_kind(change_num: usize) -> ConfChangeKind {
+        match change_num {
+            0 => ConfChangeKind::LeaveJoint,
+            1 => ConfChangeKind::Simple,
+            _ => ConfChangeKind::EnterJoint,
+        }
+    }
+}
+
+/// Abstracts over ChangePeerV2Request and (legacy) ChangePeerRequest to allow
+/// treating them in a unified manner.
+pub trait ChangePeerI {
+    type CC: ConfChangeI;
+    type CP: AsRef<[ChangePeerRequest]>;
+
+    fn get_change_peers(&self) -> Self::CP;
+
+    fn to_confchange(&self, _: Vec<u8>) -> Self::CC;
+}
+
+impl<'a> ChangePeerI for &'a ChangePeerRequest {
+    type CC = eraftpb::ConfChange;
+    type CP = Vec<ChangePeerRequest>;
+
+    fn get_change_peers(&self) -> Vec<ChangePeerRequest> {
+        vec![ChangePeerRequest::clone(self)]
+    }
+
+    fn to_confchange(&self, ctx: Vec<u8>) -> eraftpb::ConfChange {
+        let mut cc = eraftpb::ConfChange::default();
+        cc.set_change_type(self.get_change_type());
+        cc.set_node_id(self.get_peer().get_id());
+        cc.set_context(ctx);
+        cc
+    }
+}
+
+impl<'a> ChangePeerI for &'a ChangePeerV2Request {
+    type CC = eraftpb::ConfChangeV2;
+    type CP = &'a [ChangePeerRequest];
+
+    fn get_change_peers(&self) -> &'a [ChangePeerRequest] {
+        self.get_changes()
+    }
+
+    fn to_confchange(&self, ctx: Vec<u8>) -> eraftpb::ConfChangeV2 {
+        let mut cc = eraftpb::ConfChangeV2::default();
+        let changes: Vec<_> = self
+            .get_changes()
+            .iter()
+            .map(|c| {
+                let mut ccs = eraftpb::ConfChangeSingle::default();
+                ccs.set_change_type(c.get_change_type());
+                ccs.set_node_id(c.get_peer().get_id());
+                ccs
+            })
+            .collect();
+
+        if changes.len() <= 1 {
+            // Leave joint or simple confchange
```

Comment:
Why not always use `Explicit`?

Reply:
We don't want to enter joint for leave joint request or simple confchange request, the use of `ConfChangeTransition::Auto` here is subject to how `raft-rs` decide to `leave_joint`, `enter_joint` or `simple`.